### PR TITLE
Refactor std::string to std::vector/array/span for storing bytes

### PR DESCRIPTION
**Pull request by itamarcps**
This is a major update where we have comprehensively overhauled the way we handle raw byte storage in the codebase. We have shifted from using std::string and std::string_view to std::vector<uint8_t>, std::array<uint8_t> and std::span<uint8_t>. This change is motivated by a desire to enhance clarity, efficiency, and robustness in handling raw bytes, especially in the context of TCP sockets which operate on uint8_t arrays.

Our new byte storage and viewing approach employs the following:

```cpp
using Byte = uint8_t;
using Bytes = std::vector<Byte>;
template <std::size_t N>
using BytesArr = std::array<Byte, N>;
using BytesArrView = std::span<const Byte, std::dynamic_extent>;
using BytesArrMutableView = std::span<Byte, std::dynamic_extent>;
```

The breadth of the modification impacts almost every aspect of our code. Alongside the primary change, we've also introduced several enhancements and adjustments to provide better context, such as duplicating some ABI functions to accommodate both Bytes and std::string, refactoring FixedString to FixedArr and opting for std::array instead (take note that std::array cannot leverage std::move as uint8_t is not movable), and incorporating new utility functions. Numerous functions now also take std::span as an argument. I would kindly ask reviewers to pay special attention to the utils.h and strings.h classes.

The most critical alteration concerns ethCallInfo, which now includes an additional Functor variable in its tuple. Moreover, the data attribute of ethCallInfo now returns a BytesArrView, necessitating careful consideration of data ownership. The ethCallInfo structure is now as follows:

```
/// ethCallInfo: tuple of (from, to, gasLimit, gasPrice, value, functor, data)
/// **ATTENTION**: Be aware that we are using BytesArrView, so you MUST be sure that the data allocated in the BytesArrView is valid during the whole life of the tuple.
/// If you need ethCallInfo to own the data, use ethCallInfoAllocated instead.
using ethCallInfo = std::tuple<Address,Address,uint256_t, uint256_t, uint256_t, Functor, BytesArrView>;
/// This is the same as ethCallInfo, but using Bytes instead of BytesArrView, truly allocating the data, some places need it such as tests.
using ethCallInfoAllocated = std::tuple<Address,Address,uint256_t, uint256_t, uint256_t, Functor, Bytes>;
```

When using std::span, it's essential to remember that it does not own the data. This means that you need to ensure the integrity of the array that std::span is referencing for the span's duration, ethCallInfoAllocated is used when needed (such as tests). To append the return of a function to bytes, it's recommended to use Utils::appendBytes.

Given the extent of the changes, we acknowledge the possibility of conflicts, and I take full responsibility for their resolution. Your detailed review and feedback on this extensive refactor will be highly appreciated.

### CODE REVIEW

**Comment by jcarraror** [comment]
There's a lot of magic numbers in abi.h (for offsets and byte skipping). I get the context and it's clear for me, but it's not better to have this values in a proper named variable? This way we avoid any kind of confusion in interpreting that values.

**Comment by itamarcps** [comment]
> There's a lot of magic numbers in the project(for offsets and byte skipping). I get the context and it's clear for me, but it's not better to have this values in a proper named variable? This way we avoid any kind of confusion in interpreting that values.


Some might be able to be added as global const, It could be used on Block. others are "dynamic" that depends on the size of the previous, what is your suggestion?

**Comment by jcarraror** [comment]
> > There's a lot of magic numbers in the project(for offsets and byte skipping). I get the context and it's clear for me, but it's not better to have this values in a proper named variable? This way we avoid any kind of confusion in interpreting that values.
> 
> Some might be able to be added as global const, It could be used on Block. others are "dynamic" that depends on the size of the previous, what is your suggestion?

A good way is to define a header for this constants (probably in the utils folder)

**Review by jcarraror — CHANGES_REQUESTED** [review]

File: tests/contract/erc20wrapper.cpp [hunk 1]
```diff
@@ -9,7 +9,7 @@
 #include <filesystem>
 
 /// Forward Decleration.
```
Comment (Jean-Lessa) [inline tests/contract/erc20wrapper.cpp:11]:
```suggestion
// Forward declaration.
```

File: tests/contract/nativewrapper.cpp [hunk 2]
```diff
@@ -9,7 +9,8 @@
 #include "../../src/core/rdpos.h"
 
 /// Forward Decleration.
```
Comment (Jean-Lessa) [inline tests/contract/nativewrapper.cpp:11]:
```suggestion
// Forward declaration.
```

File: src/contract/variables/safeunorderedmap.h [hunk 3]
```diff
@@ -17,478 +17,478 @@
...
-        if (itM != map.end()) {
-          auto itD = erasedKeys->find(key);
-          if (itD == erasedKeys->end()) (*mapPtr)[key] = itM->second;
-        }
+private:
+  std::unordered_map<Key, T, SafeHash> map; ///< Value.
+  mutable std::unique_ptr<std::unordered_map<Key, T, SafeHash>> mapPtr; ///< Pointer to the value.
+  mutable std::unique_ptr<std::unordered_set<Key, SafeHash>> erasedKeys; ///< Pointer to the set of erased keys.
+
+  ///< Check if pointers are initialized (and initialize them if not).
```
Comment (Jean-Lessa) [inline src/contract/variables/safeunorderedmap.h:25]:
```suggestion
  /// Check if pointers are initialized (and initialize them if not).
```

File: src/contract/variables/safeunorderedmap.h [hunk 4]
```diff
@@ -17,478 +17,478 @@
...
+  ) : SafeBase(owner), map(map) {}
+
+  /**
+   * Empty constructor.
+   * @param map The initial value. Defaults to an empty map.
+   */
+  SafeUnorderedMap(const std::unordered_map<Key, T, SafeHash>& map = {})
+    : SafeBase(nullptr), mapPtr(std::make_unique<std::unordered_map<Key, T, SafeHash>>(map)) {}
+
+  ///< Copy constructor.
```
Comment (Jean-Lessa) [inline src/contract/variables/safeunorderedmap.h:109]:
```suggestion
  /// Copy constructor.
```

File: src/contract/variables/safeunorderedmap.h [hunk 5]
```diff
@@ -17,478 +17,478 @@
...
+  void commit() override {
+    check();
+    map.merge(*mapPtr);
+    for (const auto &[key, value] : (*mapPtr)) map[key] = value;
+    for (const auto& key : (*erasedKeys)) map.erase(key);
+    mapPtr = nullptr;
+    registered = false;
+  }
+
+  ///< Revert the value. Nullifies the pointers and unregisters the variable.
```
Comment (Jean-Lessa) [inline src/contract/variables/safeunorderedmap.h:158]:
```suggestion
  /// Revert the value. Nullifies the pointers and unregisters the variable.
```

File: src/contract/contractmanager.cpp [hunk 6]
```diff
@@ -52,200 +53,193 @@ Address ContractManager::deriveContractAddress(const ethCallInfo& callInfo) cons
...
+  }
 
   // Create the contract
-  this->contracts.insert(std::make_pair(derivedContractAddress, std::make_unique<ERC20>(
-    decoder.getData<std::string>(0), decoder.getData<std::string>(1),
-    uint8_t(decoder.getData<uint256_t>(2)), decoder.getData<uint256_t>(3),
-    this->interface, derivedContractAddress, this->getCaller(),
-    this->options->getChainID(), this->db
-  )));
+  this->contracts.insert(std::make_pair(derivedContractAddress, std::make_unique<ERC20>(decoder.getData<std::string>(0),
```
Comment (Jean-Lessa) [inline src/contract/contractmanager.cpp:90]: Indentation could be fixed here.

**Review by Jean-Lessa — CHANGES_REQUESTED** [review]
Finished reviewing, left a lot of small suggestions (most of them to do with either Doxygen or indentation/code formatting).

**Review by Jean-Lessa — CHANGES_REQUESTED** [review]
Finished reviewing, left a lot of small suggestions (most of them to do with either Doxygen or indentation/code formatting).

**Review by jcarraror — APPROVED** [review]

File: src/utils/db.h [hunk 7]
```diff
@@ -34,30 +34,96 @@ struct DBServer {
    * @param version The database's version string.
    */
   DBServer(std::string host, std::string version) : host(host), version(version) {};
+  const Bytes blocks = { 0x00, 0x01 };
```
[the remark this answers is not in the document]
Comment (itamarcps) [reply src/utils/db.h:37]: Yes, duplicated during the merge, removed.

File: src/utils/utils.h [hunk 8]
```diff
@@ -50,7 +50,10 @@ using BytesArr = std::array<Byte, N>;
 using BytesArrView = std::span<const Byte, std::dynamic_extent>;
 using BytesArrMutableView = std::span<Byte, std::dynamic_extent>;
 
-/// ethCallInfo: tuple of (from, to, gasLimit, gasPrice, value, functor, data)
+/**
+ * ethCallInfo: tuple of (from, to, gasLimit, gasPrice, value, functor, data)
+ * ...
```
Comment (Jean-Lessa) [inline src/utils/utils.h]: I think I specified this poorly now that I realize it, the `...` was meant to be "the rest of the comment lines below it". My bad, just fixed it myself with 998c843.

**Review by Jean-Lessa — APPROVED** [review]
All good! Approving it :ok_hand:

**Pull request merged**